x/twap/store.go

```diff
@@ -117,6 +133,24 @@ func (k Keeper) pruneRecordsBeforeTimeButNewest(ctx sdk.Context, lastKeptTime ti
 		store.Delete(timeIndexKey)
 		poolIndexKey := types.FormatHistoricalPoolIndexTWAPKeyFromStrTime(poolId, asset0, asset1, timeS)
 		store.Delete(poolIndexKey)
+
+		// Increment the number of records pruned by 2, since we delete two records per iteration.
+		numPruned += 2
+
+		if numPruned >= NumRecordsToPrunePerBlock {
+			// We have hit the limit, so we stop pruning.
+			break
+		}
+	}
+
+	if !iter.Valid() {
+		// The iterator is exhausted, so we have pruned all records.
+		state.IsPruning = false
+		k.SetPruningState(ctx, state)
+	} else {
+		// We have not pruned all records, so we update the last key seen.
+		state.LastKeySeen = iter.Key()
+		k.SetPruningState(ctx, state)
```

Comment:
I may ahve missed it, but I don't think we have any logic testing this code path. (or rather that keys from this are working as we want)

From reading the code, this loosk right to me!

Reply:
We don't, I will add before merge

Reply:
This test shows pruning happening over multiple blocks https://github.com/osmosis-labs/osmosis/pull/7427/commits/f84a716c5d272f373ef86031360b7a01e7fa6dc8

It verifies
1. We see the extra record that we save due to the bug
2. Pruning eventually gets set to false when we iterate over records
3. The correct records get pruned

@ValarDragon